### PR TITLE
Add: DeepSeek V4 Flash real-weight loading and numeric validation

### PR DESCRIPTION
**Pull request by lwDavid**
- weights_flash.py converts the hybrid MXFP4-MXFP8 checkpoint to the
  kernel ABI: FP4/FP8 block dequant with UE8M0 scales, fixture-equivalent
  per-output-channel INT8 requant, per-layer stacking with CSA/HCA kind
  slots, EP expert sharding and TP vocab sharding; ships an offline
  conversion CLI plus whole-net and single-layer spec injection
- --weights on decode_fwd/prefill_fwd loads a converted cache or the raw
  checkpoint; on decode_layer/prefill_layer it injects one layer's real
  weights so the existing layer goldens validate on real dynamic ranges
- golden_fwd.py chains embed, 43 per-layer attention+MoE goldens,
  hc_head, final norm, and LM head into a full-network prefill golden;
  prefill_fwd --validate wires it with cosine/rel-L2 logits gates
- golden runner skips the input snapshot clone when no golden_fn is set,
  halving peak host RAM for full-model-weight runs
- EP<8 loads reduce the expert space to 32*EP (first checkpoint experts,
  reduced router tables) mirroring moe.py; only EP8 is the full model

Based on #956 (deepseek_v4_pro Flash preset); merge after it.

### CODE REVIEW

**Comment by lwDavid** [comment]
Appended four commits on top of 9d97a24, taking the branch from real-weight loading/validation to a working end-to-end generation path:

- **Add: mapped-row and prefix comparators for full-net validation** — cache pools are compared per layer/rank on mapped active rows only (no dilution by unmapped slots), hidden states honor the runtime `num_tokens` with a bit-exact tail, and sampled ids must match the torch golden argmax.
- **Fix: deterministic split-K reductions and A5-exact goldens** — atomic-add split-K reductions replaced by ordered partial reduction; goldens model the A5 TROWSUM tree, Cube K16 MAD group rounding, and high-precision rsqrt.
- **Fix: split SWA and compressed RoPE profiles and retire MoE credits** — SWA layers no longer share the compressed layers' RoPE table; persistent MoE signal windows are retired per dispatch (trace-time escape hatch below).
- **Add: real-weight prompt generation in the token loop** — prompt → tokenizer → resident prefill + greedy decode session → detokenized text, with per-step cross-rank agreement asserts.

Validation on an 8×Ascend950PR host against the released DeepSeek-V4-Flash checkpoint:

- torch golden (real prompt, replicated ranks): all ranks bit-identical, argmax `" Paris"` at EP2 and EP8
- device vs golden under the strict comparators (EP2 and EP8): `logits` PASS (cosine ≥ 0.99) and `sampled_ids` PASS against the golden argmax; per-element cache-pool gates still exceed their budgets from a ~1% first-CSA-layer residual that accumulates over depth — gate numbers are unchanged, the comparators just no longer dilute
- end-to-end EP8 greedy generation (32 steps): `" Paris. It is located in the north-central part of the country, on the Seine River. Paris is known for its art, fashion, gastronomy, and culture"`, all 8 ranks agreeing at every step
- host suites: `pytest tests/golden` 246 passed; the seven new model tests pass alone and combined; `pre-commit run --all-files` clean

Two open EP8 issues are documented in `docs/models/deepseek_v4_pro.md`, both with workarounds wired into the recipe:

1. `moe_signal_retire` can deadlock an EP8 dispatch: the single-element anchor orders the negative credits only after the task writing `pre_hc_hidden_out[0,0,0]`, so they can land while later waits of the same dispatch are still pending. `DSV4_DISABLE_MOE_RETIRE=1` compiles a forward without the scope — use it for prefill; multi-step decode keeps retirement (without it, decode logits go non-finite on part of the ranks by the second step).
2. An EP8 prefill compiled at `--num-tokens 128` stalls on-device (`S1:running-stalled`, the same task id on every rank); the same source at `--num-tokens 16` runs. EP2@128 and EP8@6/16 are fine.

File: docs/models/deepseek_v4_pro.md [hunk 1]
```diff
@@ -122,6 +122,91 @@ prefill_mtp     mtp_projection → prefill_attention_swa → moe → hc_head →
...
+  trace-time knob `DSV4_DISABLE_MOE_RETIRE=1` compiles a forward without
+  the scope. For the token loop, compile prefill with the knob set and
+  decode without it: prefill's un-retired credits did not disturb the
+  following decode steps in testing, while multi-step decode does require
+  its own retirement (without it, decode produced non-finite logits on part
+  of the ranks by the second step).
+- An EP8 prefill compiled at `--num-tokens 128` stalls on-device
+  (`S1:running-stalled` on the same task id on every rank); the same source
+  at `--num-tokens 16` runs. Keep the prefill MoE extent small until the
+  stall is root-caused (EP2 at 128 and EP8 at 6/16 both run).
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply docs/models/deepseek_v4_pro.md]: Fixed — the loop now has per-phase compile control: `--prefill-tokens N` sets the prefill MoE extent and `--prefill-no-retire` injects `DSV4_DISABLE_MOE_RETIRE=1` into the prefill compile child only (one note: `_compile_program` copies the whole parent env, so the knob did propagate — the gap was per-phase control, which the flag now provides). The EP8 example carries `--prefill-tokens 16 --prefill-no-retire` and runs as written; both caveats now reference the flags. Verified by compiling the doc example: the prefill artifact contains no `moe_signal_retire` scope while the decode artifact keeps it.


Fixed in b6ef734.

File: models/deepseek_v4_pro/golden_fwd.py [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+                if int(row_indices[rank, row]) < 0:
+                    continue
+                a_row = a[rank, row]
+                e_row = e[rank, row]
+                if not bool(torch.isfinite(a_row).all()):
+                    failures.append(f"    logits[{rank},{row}]: non-finite values")
+                    continue
+                denom = float(a_row.norm() * e_row.norm())
+                cosine = float(a_row @ e_row) / denom if denom > 0.0 else 0.0
+                rel_l2 = float((a_row - e_row).norm() / e_row.norm().clamp_min(1e-12))
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/deepseek_v4_pro/golden_fwd.py]: Fixed as suggested — the comparator now rejects rows whose expected side is non-finite, with distinct "non-finite actual/expected values" messages.


Fixed in b6ef734.

File: models/deepseek_v4_pro/prefill_fwd.py [hunk 3]
```diff
@@ -1558,24 +1687,99 @@ def main():
...
+            raise SystemExit("--prompt-file requires --tokenizer")
+        tokenizer = Tokenizer.from_file(args.tokenizer)
+        prompt_ids = tokenizer.encode(Path(args.prompt_file).read_text()).ids
+        bos_id = tokenizer.token_to_id("<｜begin▁of▁sentence｜>")
+        if bos_id is not None:
+            prompt_ids = [bos_id] + prompt_ids
+        if not 1 <= len(prompt_ids) <= T:
+            raise SystemExit(f"prompt is {len(prompt_ids)} tokens; must be within 1..{T}")
+        args.num_tokens = len(prompt_ids)
+        print(f"[RUN] prompt: {len(prompt_ids)} tokens {prompt_ids}", flush=True)
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/deepseek_v4_pro/prefill_fwd.py:1740]: The repo ships no packaging metadata (it is consumed from the checkout via PYTHONPATH), so there is no dependency list to declare `tokenizers` in; took the fallback: all three `--prompt*`/`--tokenizer` import sites now catch the ImportError and exit with a "pip install tokenizers" hint. The fullwidth `｜` RUF001 hits are intentional — that is the literal BOS token string of the checkpoint's tokenizer.


Fixed in b6ef734.

File: models/deepseek_v4_pro/qkv_proj_rope.py [hunk 4]
```diff
@@ -519,33 +702,46 @@ def apply_rope(x_rope, cos, sin):
...
+    kv_full = rms_norm(
+        split_k_matmul_bf16_input_fp32(
+            token_x,
+            wkv,
+            splits=KV_OK,
+            k_per_split=KV_K_SLICE,
+            k_tile=KV_K_TILE,
+        ),
+        gamma_ckv,
+    )  # [T, HEAD_DIM]
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/deepseek_v4_pro/qkv_proj_rope.py:718]: Aligned as proposed — the KV path now uses `_golden_a5_split_k_bf16_matmul`; `kv_proj_matmul` is indeed the same construct as `qr_proj_matmul` (BF16 inputs, one FP32 Cube accumulator per split via matmul_acc, ordered split reduction), and `KV_K_SLICE` is K16-divisible so the contiguous-K16 model applies unchanged. The now-unused local `split_k_matmul_bf16_input_fp32` helper is removed. `tests/golden` and the model host suites pass against the tightened reference.


Fixed in b6ef734.

File: tests/models/test_deepseek_v4_hc_pre_rms.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+def test_a5_high_precision_rsqrt_avoids_fp32_sqrt_reciprocal_double_rounding():
+    # This exact hc_pre RMS argument was observed at Flash prefill token 4.
+    # The A5 high-precision path returns the correctly rounded FP32 reciprocal
+    # square root; host FP32 sqrt/reciprocal lands one ULP lower.
+    rms_arg = torch.tensor([0x3B99BBDE], dtype=torch.int32).view(torch.float32)
+
+    actual = _golden_a5_high_precision_rsqrt(rms_arg)
+    assert actual.view(torch.int32).item() == 0x416997DC
+    assert torch.rsqrt(rms_arg).view(torch.int32).item() == 0x416997DB
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply tests/models/test_deepseek_v4_hc_pre_rms.py]: Agreed on the portability concern. Kept the golden-side exact assertion (the golden rounds once through float64, so its bit pattern is host-independent) and dropped the host `torch.rsqrt` assertion entirely rather than inverting it — the `!=` form has the mirrored problem on a host whose rsqrt happens to be correctly rounded. The observed one-ULP-low host value stays as a comment.


Fixed in b6ef734.

File: golden/__init__.py [hunk 6]
```diff
@@ -46,9 +46,12 @@ def _configure_golden_threads() -> int:
 from .spec import ScalarSpec, TensorSpec
 from .validation import (
     error_distribution,
+    input_prefix_ratio_allclose,
     mapped_pool_ratio_allclose,
     ratio_allclose,
     ratio_reldiff,
+    sampled_ids_golden_compare,
+    stacked_mapped_pool_ratio_allclose,
```
Comment (zhangqi-chen) [inline golden/__init__.py]: Check if the 3 new validate functions are necessary.

File: tests/models/test_deepseek_v4_decode_rope_profiles.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
+# Copyright (c) PyPTO Contributors.
```
Comment (zhangqi-chen) [inline tests/models/test_deepseek_v4_decode_rope_profiles.py]: delete tests/models

**Review by zhangqi-chen — CHANGES_REQUESTED** [review]

File: golden/__init__.py (hunk 6 above)
Comment (lwDavid) [reply golden/__init__.py]: Done in 4ad1465 — the golden package is reverted to its pre-PR state. The three comparators now live in `models/deepseek_v4_pro/utils/golden_fwd.py` next to their only consumer; they are model-specific by design (stacked pool layout, prompt-prefix semantics, lm_head sampling) and internally compose the existing public `ratio_allclose` primitive rather than duplicating it. The only remaining public change is the generic effective-scalar pass-through in `golden/runner.py`, which any comparator needs to see replayed scalar inputs.

File: tests/models/test_deepseek_v4_decode_rope_profiles.py (hunk 7 above)
Comment (lwDavid) [reply tests/models/test_deepseek_v4_decode_rope_profiles.py]: Done in 4ad1465 — `tests/models/` is removed entirely (all eight files, including the earlier quantization-helper tests). The only surviving test change is a one-line fix in the pre-existing `tests/golden/test_validation.py` suite that derives an assertion threshold from `Q_LORA` instead of hardcoding the pro-variant value, so the suite passes regardless of which variant the driver modules were imported with.

**Comment by lwDavid** [comment]
Addressed the structural review in 4ad1465 (all four points):

1. **Shared `golden/` kept minimal** — `golden/__init__.py` and `golden/validation.py` are reverted to their pre-PR state. The three full-net comparators move to `models/deepseek_v4_pro/utils/golden_fwd.py` beside their only consumer; they compose the existing public `ratio_allclose` primitive internally. The one remaining shared change is the generic effective-scalar pass-through in `golden/runner.py`.
2. **`weights_flash.py` → `models/deepseek_v4_pro/utils/`** — it is a loader, not an operator. The conversion CLI still runs as a plain script (`python models/deepseek_v4_pro/utils/weights_flash.py ...`); docs updated.
3. **`golden_fwd.py` → `models/deepseek_v4_pro/utils/`, goldens pushed into operator files** — the leaf goldens always lived in their operator files; this file is the composition of them in `prefill_fwd`'s layer schedule plus the compare-fn builder, which is why it exists. Its two operator-level rewrites are now sunk into their homes: `hc_head.golden_hc_head_rows` (row-count-agnostic) and `lm_head.golden_lm_head_all_ranks(n_ranks=...)`, so `utils/golden_fwd.py` keeps orchestration only.
4. **`tests/models/` deleted** — all eight files.

Behavior is unchanged: `pytest tests/golden` passes (233), `pre-commit run --all-files` is clean, the conversion CLI runs from its new path, and an EP2 real-weight `--validate --save-data` run through the relocated modules reproduces the golden anchor exactly (argmax `" Paris"`, ranks bit-identical).

**Review by zhangqi-chen — APPROVED** [review]
